File: src/main/java/com/example/finsentinel/service/scraper/FirecrawlClient.java
package com.example.finsentinel.service.scraper;

import com.example.finsentinel.config.FirecrawlProperties;
import tools.jackson.databind.JsonNode;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

import java.util.Map;

/**
 * Implements firecrawl client business operations and integrations.
 *
 * <p>This class is part of the service layer in FinSentinel.
 */

@Service
@Slf4j
public class FirecrawlClient {

    private final RestClient restClient;
    private final FirecrawlProperties properties;

    /**
     * Creates a new FirecrawlClient instance.
     *
     * <p>This method is defined in {@link FirecrawlClient}.
     * @param restClientBuilder rest client builder (RestClient.Builder)
     * @param properties properties (FirecrawlProperties)
     */

    public FirecrawlClient(RestClient.Builder restClientBuilder, FirecrawlProperties properties) {
        this.properties = properties;
        this.restClient = restClientBuilder
                .baseUrl(properties.getBaseUrl())
                .defaultHeader("Authorization", "Bearer " + properties.getApiKey())
                .build();
    }

    private static final int MAX_RETRIES = 3;
    private static final long INITIAL_BACKOFF_MS = 1000;

    /**
     * Scrape a single URL and return the markdown content.
     * Returns null if scraping fails after all retries.
     */
    public ScrapeResult scrape(String url) {
        for (int attempt = 1; attempt <= MAX_RETRIES; attempt++) {
            try {
                JsonNode response = restClient.post()
                        .uri("/scrape")
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(Map.of(
                                "url", url,
                                "formats", new String[]{"markdown"}
                        ))
                        .retrieve()
                        .body(JsonNode.class);

                if (response != null && response.has("data")) {
                    JsonNode data = response.get("data");
                    String markdown = data.has("markdown") ? data.get("markdown").asText() : "";
                    String title = data.has("metadata") && data.get("metadata").has("title")
                            ? data.get("metadata").get("title").asText()
                            : url;
                    return new ScrapeResult(title, markdown, url);
                }
                log.warn("Firecrawl returned no data for URL (attempt {}/{}): {}", attempt, MAX_RETRIES, url);
            } catch (Exception e) {
                log.warn("Firecrawl scrape failed (attempt {}/{}): {} - {}", attempt, MAX_RETRIES, url, e.getMessage());
                if (attempt < MAX_RETRIES) {
                    try {
                        long backoffMs = INITIAL_BACKOFF_MS * (1L << (attempt - 1));
                        Thread.sleep(backoffMs);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        return null;
                    }
                }
            }
        }
        log.error("Firecrawl scrape failed after {} retries for URL: {}", MAX_RETRIES, url);
        return null;
    }

    /**
     * Implements scrape result business operations and integrations.
     *
     * <p>This record belongs to the service layer in FinSentinel.
     */

    public record ScrapeResult(String title, String markdown, String sourceUrl) {}
}

File: apps/api/src/scraper/firecrawl.client.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

interface FirecrawlResponse {
  success: boolean;
  data?: {
    markdown?: string;
  };
  error?: string;
}

/**
 * REST client for the Firecrawl web scraping API.
 *
 * POST {baseUrl}/scrape with Bearer token auth.
 * Retry logic: 3 attempts with 1-second backoff.
 */
@Injectable()
export class FirecrawlClient {
  private readonly logger = new Logger(FirecrawlClient.name);
  private readonly apiKey: string;
  private readonly baseUrl: string;

  private static readonly MAX_RETRIES = 3;
  private static readonly RETRY_DELAY_MS = 1_000;

  constructor(configService: ConfigService) {
    this.apiKey = configService.get<string>('firecrawl.apiKey', '');
    this.baseUrl = configService.get<string>('firecrawl.baseUrl', 'https://api.firecrawl.dev/v2');
  }

  /**
   * Scrape a URL and return its content as markdown.
   * Returns null if the response contains no markdown data.
   * Throws if all retry attempts are exhausted.
   */
  async scrape(url: string): Promise<string | null> {
    let lastError: Error | null = null;

    for (let attempt = 1; attempt <= FirecrawlClient.MAX_RETRIES; attempt++) {
      try {
        const response = await fetch(`${this.baseUrl}/scrape`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            Authorization: `Bearer ${this.apiKey}`,
          },
          body: JSON.stringify({
            url,
            formats: ['markdown'],
          }),
        });

        if (!response.ok) {
          const body = (await response.json()) as FirecrawlResponse;
          lastError = new Error(
            `Firecrawl HTTP ${response.status}: ${body.error ?? response.statusText}`,
          );
          this.logger.warn(
            `Firecrawl attempt ${attempt}/${FirecrawlClient.MAX_RETRIES} failed: ${lastError.message}`,
          );

          if (attempt < FirecrawlClient.MAX_RETRIES) {
            await this.sleep(FirecrawlClient.RETRY_DELAY_MS);
          }
          continue;
        }

        const body = (await response.json()) as FirecrawlResponse;
        return body.data?.markdown ?? null;
      } catch (err) {
        lastError = err instanceof Error ? err : new Error(String(err));
        this.logger.warn(
          `Firecrawl attempt ${attempt}/${FirecrawlClient.MAX_RETRIES} error: ${lastError.message}`,
        );

        if (attempt < FirecrawlClient.MAX_RETRIES) {
          await this.sleep(FirecrawlClient.RETRY_DELAY_MS);
        }
      }
    }

    throw new Error(
      `Firecrawl scrape failed after ${FirecrawlClient.MAX_RETRIES} attempts for ${url}: ${lastError?.message}`,
    );
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}
